### DISCUSSION TITLE
Updating an async atom twice from a writer atom never resolves

### DISCUSSION THREAD

**Discussion by martinhath** [pmndrs/jotai#2682] [Bug report]
### Bug Description

I've found what I think is a bug and have written two test cases that are seemingly equivalent, but where one succeeds and one fails. 

The succeding one creates an `atom<Promise<number>>` and a writer atom that increments the promised value twice. All is well.

```tsx
it('write to async atom twice', async () => {
  const asyncAtom = atom(Promise.resolve(2))

  const writer = atom(null, async (get, set) => {
    let val = await get(asyncAtom)
    set(asyncAtom, Promise.resolve(val + 1))
    val = await get(asyncAtom)
    set(asyncAtom, Promise.resolve(val + 1))
    return get(asyncAtom)
  })

  const Component = () => {
    const load = useAtomValue(loadable(asyncAtom))
    const write = useSetAtom(writer)
    const val = load.state === 'hasData' ? load.data : 'loading'
    return (
      <>
        <div>atom count: {val}</div>
        <button onClick={() => write()}>dispatch</button>
      </>
    )
  }

  const { findByText, getByText } = render(
    <StrictMode>
      <Component />
    </StrictMode>,
  )

  await findByText('atom count: 2')

  await userEvent.click(getByText('dispatch'))
  await findByText('atom count: 4', undefined, { timeout: 1000 })
})
```

However, if we change the `writer` to use an update function instead:
```ts
  const writer = atom(null, (get, set) => {
    set(asyncAtom, async (c) => (await c) + 1)
    set(asyncAtom, async (c) => (await c) + 1) // This await gets stuck
    return get(asyncAtom)
  })
```
it does not work.  Some `console.log` debugging shows that the returned promise doesn't resolve, and that the second `await c` doesn't resolve. 

### Reproduction Link

https://codesandbox.io/p/sandbox/serene-paper-v9n5qy?file=%2Fsrc%2FApp.tsx

**Comment by dai-shi** [comment]
Thanks for reporting!
Can you try if previous versions have worked before? Especially, v2.8.4.

**Reply by martinhath** [reply]
I quickly ran through all `2.x.0` versions (and also `2.8.4`) on codesandbox.io, and it seems that the only version that works is `2.0`; `2.1` is broken. All patch versions `2.0.x` work.

**Reply by dai-shi** [reply]
Thanks. That's very old. I think it worked in v2.0.x by chance.

**Comment by dai-shi** [comment]
Reproduction with StackBlitz: https://stackblitz.com/edit/vitejs-vite-indg9c?file=src%2FApp.tsx

If we add a setTimeout wait, it works. This is a bug, but I'm not sure how easy a fix would be. Would anyone like to try?

**Reply by dai-shi** [reply]
a failing test: #2695

**Reply by dai-shi** [reply]
okay, it's like a deep issue. I'll work on it.

**Reply by dai-shi** [reply]
I'm still investigating it, but it may require a big architectural change. If that's the case, it will be a minor bump and may take time.

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/pull/2695 is done. Can you try it please?

https://stackblitz.com/edit/vitejs-vite-rhna75?file=package.json seems to work.

**Reply by martinhath** [reply]
That's great! It seems like promises are no longer stuck, but I did hit this error immediately:
```
Uncaught (in promise) Error: [Bug] p is not updated even after cancelation
```

I can see if I can get a small reproduction; this is from a 300kloc app 😅
